File: bindings/rust/zova/src/kv.rs
use crate::error::Result;
use std::ptr;

/// Borrowed key-value pair for a batch key-value put.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct KvEntry<'a> {
    pub key: &'a [u8],
    pub value: &'a [u8],
}

impl<'a> KvEntry<'a> {
    pub fn new(key: &'a [u8], value: &'a [u8]) -> Self {
        Self { key, value }
    }
}

fn raw_bytes(data: &[u8]) -> zova_sys::zova_kv_bytes {
    zova_sys::zova_kv_bytes {
        data: data.as_ptr(),
        len: data.len(),
    }
}

pub(crate) fn kv_get_raw(
    db: *mut zova_sys::zova_database,
    status: impl FnOnce(i32) -> Result<()>,
    namespace: &[u8],
    key: &[u8],
) -> Result<Option<Vec<u8>>> {
    let mut result = zova_sys::zova_kv_get_result {
        found: 0,
        value: zova_sys::zova_buffer {
            data: ptr::null_mut(),
            len: 0,
        },
    };
    let request = zova_sys::zova_kv_get_request {
        db,
        ns: raw_bytes(namespace),
        key: raw_bytes(key),
        out_result: &mut result,
    };
    status(unsafe { zova_sys::zova_kv_get(&request) })?;
    let found = result.found != 0;
    let value = read_owned_buffer(result.value);
    unsafe { zova_sys::zova_kv_get_result_free(&mut result) };
    if found {
        Ok(Some(value))
    } else {
        Ok(None)
    }
}

pub(crate) fn kv_get_many_raw(
    db: *mut zova_sys::zova_database,
    status: impl FnOnce(i32) -> Result<()>,
    namespace: &[u8],
    keys: &[&[u8]],
) -> Result<Vec<Option<Vec<u8>>>> {
    let raw_keys: Vec<zova_sys::zova_kv_bytes> =
        keys.iter().map(|key| raw_bytes(key)).collect();
    let mut results = zova_sys::zova_kv_get_many_results {
        items: ptr::null_mut(),
        len: 0,
    };
    let request = zova_sys::zova_kv_get_many_request {
        db,
        ns: raw_bytes(namespace),
        keys: raw_keys.as_ptr(),
        keys_len: raw_keys.len(),
        out_results: &mut results,
    };
    status(unsafe { zova_sys::zova_kv_get_many(&request) })?;
    let mut out = Vec::with_capacity(results.len);
    for index in 0..results.len {
        let item = unsafe { &*results.items.add(index) };
        let value = read_owned_buffer(item.value);
        if item.found != 0 {
            out.push(Some(value));
        } else {
            out.push(None);
        }
    }
    unsafe { zova_sys::zova_kv_get_many_results_free(&mut results) };
    Ok(out)
}

pub(crate) fn kv_put_raw(
    db: *mut zova_sys::zova_database,
    status: impl FnOnce(i32) -> Result<()>,
    namespace: &[u8],
    key: &[u8],
    value: &[u8],
) -> Result<()> {
    let request = zova_sys::zova_kv_put_request {
        db,
        ns: raw_bytes(namespace),
        key: raw_bytes(key),
        value: raw_bytes(value),
    };
    status(unsafe { zova_sys::zova_kv_put(&request) })
}

pub(crate) fn kv_put_many_raw(
    db: *mut zova_sys::zova_database,
    status: impl FnOnce(i32) -> Result<()>,
    namespace: &[u8],
    entries: &[KvEntry<'_>],
) -> Result<()> {
    let raw: Vec<zova_sys::zova_kv_put_entry> = entries
        .iter()
        .map(|entry| zova_sys::zova_kv_put_entry {
            key: raw_bytes(entry.key),
            value: raw_bytes(entry.value),
        })
        .collect();
    let request = zova_sys::zova_kv_put_many_request {
        db,
        ns: raw_bytes(namespace),
        entries: raw.as_ptr(),
        entries_len: raw.len(),
    };
    status(unsafe { zova_sys::zova_kv_put_many(&request) })
}

pub(crate) fn kv_delete_raw(
    db: *mut zova_sys::zova_database,
    status: impl FnOnce(i32) -> Result<()>,
    namespace: &[u8],
    key: &[u8],
) -> Result<()> {
    let request = zova_sys::zova_kv_delete_request {
        db,
        ns: raw_bytes(namespace),
        key: raw_bytes(key),
    };
    status(unsafe { zova_sys::zova_kv_delete(&request) })
}

pub(crate) fn kv_delete_many_raw(
    db: *mut zova_sys::zova_database,
    status: impl FnOnce(i32) -> Result<()>,
    namespace: &[u8],
    keys: &[&[u8]],
) -> Result<()> {
    let raw_keys: Vec<zova_sys::zova_kv_bytes> =
        keys.iter().map(|key| raw_bytes(key)).collect();
    let request = zova_sys::zova_kv_delete_many_request {
        db,
        ns: raw_bytes(namespace),
        keys: raw_keys.as_ptr(),
        keys_len: raw_keys.len(),
    };
    status(unsafe { zova_sys::zova_kv_delete_many(&request) })
}

pub(crate) fn kv_count_raw(
    db: *mut zova_sys::zova_database,
    status: impl FnOnce(i32) -> Result<()>,
    namespace: &[u8],
) -> Result<u64> {
    let mut count = 0;
    let request = zova_sys::zova_kv_count_request {
        db,
        ns: raw_bytes(namespace),
        out_count: &mut count,
    };
    status(unsafe { zova_sys::zova_kv_count(&request) })?;
    Ok(count)
}

pub(crate) fn kv_clear_namespace_raw(
    db: *mut zova_sys::zova_database,
    status: impl FnOnce(i32) -> Result<()>,
    namespace: &[u8],
) -> Result<()> {
    let request = zova_sys::zova_kv_clear_namespace_request {
        db,
        ns: raw_bytes(namespace),
    };
    status(unsafe { zova_sys::zova_kv_clear_namespace(&request) })
}

fn read_owned_buffer(buffer: zova_sys::zova_buffer) -> Vec<u8> {
    if buffer.data.is_null() || buffer.len == 0 {
        return Vec::new();
    }
    unsafe { std::slice::from_raw_parts(buffer.data, buffer.len) }.to_vec()
}

File: src/c_api/kv.zig
//! Key-value batch and single-operation entrypoint implementations.

const allocator = @import("values.zig").allocator;
const bytesConst = @import("values.zig").bytesConst;
const databaseHandle = @import("handles.zig").databaseHandle;
const failDb = @import("errors.zig").failDb;
const kvKeySlices = @import("values.zig").kvKeySlices;
const kvPutEntrySlices = @import("values.zig").kvPutEntrySlices;
const okDb = @import("errors.zig").okDb;
const zova_kv_clear_namespace_request = @import("types.zig").zova_kv_clear_namespace_request;
const zova_kv_count_request = @import("types.zig").zova_kv_count_request;
const zova_kv_delete_many_request = @import("types.zig").zova_kv_delete_many_request;
const zova_kv_delete_request = @import("types.zig").zova_kv_delete_request;
const zova_kv_get_many_request = @import("types.zig").zova_kv_get_many_request;
const zova_kv_get_request = @import("types.zig").zova_kv_get_request;
const zova_kv_get_result = @import("types.zig").zova_kv_get_result;
const zova_kv_put_many_request = @import("types.zig").zova_kv_put_many_request;
const zova_kv_put_request = @import("types.zig").zova_kv_put_request;
const zova_status = @import("types.zig").zova_status;

pub fn zova_kv_get(request: ?*const zova_kv_get_request) callconv(.c) zova_status {
    const req = request orelse return .INVALID_ARGUMENT;
    const handle = databaseHandle(req.db) orelse return .INVALID_ARGUMENT;
    handle.mutex.lock();
    defer handle.mutex.unlock();
    const out = req.out_result orelse return failDb(handle, error.InvalidArgument);
    out.* = .{ .found = 0, .value = .{ .data = null, .len = 0 } };

    const namespace = bytesConst(req.ns.data, req.ns.len) orelse return failDb(handle, error.InvalidArgument);
    const key = bytesConst(req.key.data, req.key.len) orelse return failDb(handle, error.InvalidArgument);

    var result = handle.db.kvGet(allocator, namespace, key) catch |err| return failDb(handle, err);
    // Transfer ownership of the allocation from kv_impl.GetResult to zova_kv_get_result.
    out.* = .{ .found = if (result.found) 1 else 0, .value = .{ .data = result.value.ptr, .len = result.value.len } };
    result.value = &.{};
    return okDb(handle);
}

pub fn zova_kv_get_many(request: ?*const zova_kv_get_many_request) callconv(.c) zova_status {
    const req = request orelse return .INVALID_ARGUMENT;
    const handle = databaseHandle(req.db) orelse return .INVALID_ARGUMENT;
    handle.mutex.lock();
    defer handle.mutex.unlock();
    const out = req.out_results orelse return failDb(handle, error.InvalidArgument);
    out.* = .{ .items = null, .len = 0 };

    const namespace = bytesConst(req.ns.data, req.ns.len) orelse return failDb(handle, error.InvalidArgument);
    const keys = kvKeySlices(req.keys, req.keys_len) catch |err| return failDb(handle, err);
    defer allocator.free(keys);

    const results = handle.db.kvGetMany(allocator, namespace, keys) catch |err| return failDb(handle, err);
    errdefer {
        for (results) |item| item.deinit(allocator);
        allocator.free(results);
    }

    const items = allocator.alloc(zova_kv_get_result, results.len) catch |err| return failDb(handle, err);
    for (results, items) |*result, *item| {
        item.* = .{ .found = if (result.found) 1 else 0, .value = .{ .data = result.value.ptr, .len = result.value.len } };
        result.value = &.{};
    }
    allocator.free(results);
    out.* = .{ .items = items.ptr, .len = items.len };
    return okDb(handle);
}

pub fn zova_kv_put(request: ?*const zova_kv_put_request) callconv(.c) zova_status {
    const req = request orelse return .INVALID_ARGUMENT;
    const handle = databaseHandle(req.db) orelse return .INVALID_ARGUMENT;
    handle.mutex.lock();
    defer handle.mutex.unlock();
    const namespace = bytesConst(req.ns.data, req.ns.len) orelse return failDb(handle, error.InvalidArgument);
    const key = bytesConst(req.key.data, req.key.len) orelse return failDb(handle, error.InvalidArgument);
    const value = bytesConst(req.value.data, req.value.len) orelse return failDb(handle, error.InvalidArgument);
    handle.db.kvPut(namespace, key, value) catch |err| return failDb(handle, err);
    return okDb(handle);
}

pub fn zova_kv_put_many(request: ?*const zova_kv_put_many_request) callconv(.c) zova_status {
    const req = request orelse return .INVALID_ARGUMENT;
    const handle = databaseHandle(req.db) orelse return .INVALID_ARGUMENT;
    handle.mutex.lock();
    defer handle.mutex.unlock();
    const namespace = bytesConst(req.ns.data, req.ns.len) orelse return failDb(handle, error.InvalidArgument);
    const entries = kvPutEntrySlices(req.entries, req.entries_len) catch |err| return failDb(handle, err);
    defer allocator.free(entries);
    handle.db.kvPutMany(namespace, entries) catch |err| return failDb(handle, err);
    return okDb(handle);
}

pub fn zova_kv_delete(request: ?*const zova_kv_delete_request) callconv(.c) zova_status {
    const req = request orelse return .INVALID_ARGUMENT;
    const handle = databaseHandle(req.db) orelse return .INVALID_ARGUMENT;
    handle.mutex.lock();
    defer handle.mutex.unlock();
    const namespace = bytesConst(req.ns.data, req.ns.len) orelse return failDb(handle, error.InvalidArgument);
    const key = bytesConst(req.key.data, req.key.len) orelse return failDb(handle, error.InvalidArgument);
    handle.db.kvDelete(namespace, key) catch |err| return failDb(handle, err);
    return okDb(handle);
}

pub fn zova_kv_delete_many(request: ?*const zova_kv_delete_many_request) callconv(.c) zova_status {
    const req = request orelse return .INVALID_ARGUMENT;
    const handle = databaseHandle(req.db) orelse return .INVALID_ARGUMENT;
    handle.mutex.lock();
    defer handle.mutex.unlock();
    const namespace = bytesConst(req.ns.data, req.ns.len) orelse return failDb(handle, error.InvalidArgument);
    const keys = kvKeySlices(req.keys, req.keys_len) catch |err| return failDb(handle, err);
    defer allocator.free(keys);
    handle.db.kvDeleteMany(namespace, keys) catch |err| return failDb(handle, err);
    return okDb(handle);
}

pub fn zova_kv_count(request: ?*const zova_kv_count_request) callconv(.c) zova_status {
    const req = request orelse return .INVALID_ARGUMENT;
    const handle = databaseHandle(req.db) orelse return .INVALID_ARGUMENT;
    handle.mutex.lock();
    defer handle.mutex.unlock();
    const out = req.out_count orelse return failDb(handle, error.InvalidArgument);
    const namespace = bytesConst(req.ns.data, req.ns.len) orelse return failDb(handle, error.InvalidArgument);
    out.* = handle.db.kvCount(namespace) catch |err| return failDb(handle, err);
    return okDb(handle);
}

pub fn zova_kv_clear_namespace(request: ?*const zova_kv_clear_namespace_request) callconv(.c) zova_status {
    const req = request orelse return .INVALID_ARGUMENT;
    const handle = databaseHandle(req.db) orelse return .INVALID_ARGUMENT;
    handle.mutex.lock();
    defer handle.mutex.unlock();
    const namespace = bytesConst(req.ns.data, req.ns.len) orelse return failDb(handle, error.InvalidArgument);
    handle.db.kvClearNamespace(namespace) catch |err| return failDb(handle, err);
    return okDb(handle);
}
